### ISSUE TITLE
Deep Rectifier Networks as Orthogonal Function Spaces

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#2798]
Deep Rectifier Networks as Orthogonal Function Spaces<br>
<br>
https://ift.tt/2T7w0Xp<br>
<br>
<br>
<br>
<div>
<div>
<h1>Introduction:</h1>
<p>After thinking about deep rectifier networks over the weekend it occurred to me that a function space interpretation may be useful. Briefly speaking:</p>
<ol>
<li>The ReLU activation serves as a gating mechanism for a deep network with nodes.</li>
<li>This gating mechanism decomposes the latent space of a deep rectifier network into linear feature maps .</li>
<li>Each of these feature maps have compact support and their domains are pair-wise disjoint.</li>
<li>It follows that a deep rectifier network may be considered an orthogonal space of functions under a simple coordinate transformation.</li>
<li>The continuity and compact support of the implies that the are square integrable so the latent space of forms a Hilbert space. This is interesting because many natural signals, ex. sounds, inhabit Hilbert spaces.</li>
</ol>
<p>One particular reason why I find the function space perspective useful is that I believe it will generalise quite well to deep networks used in real applications. It also leads to a natural interpretation of Bernoulli dropout as a method for maximising the effective size of this function space.</p>
<h1>Analysis of the ReLU activation:</h1>
<h2>Derivation from the sigmoid:</h2>
<p>The softplus: can be derived directly from the integral of the sigmoid:</p>
<p>\begin{equation} \int \sigma(x) dx = \ln(1+e^x) + C \end{equation}</p>
<p>so we may view the softplus as being the accumulation of input from an infinite number of perturbed sigmoids.</p>
<p>Now, from the softplus we may derive the ReLU which we abbreviate as :</p>
<p>\begin{equation} \ln(1+e^x) \approx g(x) = x \cdot (x \geq 0) \end{equation}</p>
<p>when . Mathematically, this has the effect of being a strict gating mechanism.</p>
<h2>Algebraic properties of ReLU:</h2>
<p>For any vector or matrix , or tensors in general, we may use to denote the application of to each component of or respectively. From this simple definition we note that:</p>
<h3>distributes over addition only in special circumstances:</h3>
<p>\begin{equation} \hat{g}(W + \Delta W) = \hat{g}(W) + \hat{g}(\Delta W) \end{equation}</p>
<p>if and only if and are both non-negative tensors.</p>
<h3>Iterated application of is equivalent to :</h3>
<p>\begin{equation} \hat{g}^n = Id \circ \hat{g} = \hat{g} \end{equation}</p>
<p>where is the identity mapping.</p>
<h3>Existence of the inverse :</h3>
<p>The inverse exists only when restricted to the union of sets where . Naturally, these sets must be non-negative.</p>
<h1>Orthogonal function spaces:</h1>
<h2>Mathematical definition of deep rectifier networks:</h2>
<p>A deep rectifier network is typically introduced as a composition of simple functions which defines a non-linear mapping:</p>
<p>where the parameter space is defined as follows:</p>
<p>\begin{equation} \theta = \big\{\ W_l \in \mathbb{R}^{k n_l \times n_{l-1}},b_l \in \mathbb{R}^{k n_l}: l \in [L] \big\} \end{equation}</p>
<h2>as a function space:</h2>
<p>From the definition, we may deduce that the total number of nodes in the deep rectifier network is given by:</p>
<p>\begin{equation} N = \sum_{l=1}^L n_l \end{equation}</p>
<p>and from this it follows that we may interpret a deep rectifier network as a space of at most functions since each node in may be &lsquo;on&rsquo; or &lsquo;off&rsquo;.</p>
<h2>partitions the domain of :</h2>
<p>It&rsquo;s useful to note that if the domain of is , the associated functions partition into pair-wise disjoint compact sets:</p>
<p>\begin{equation} X_i \cap X_{j \neq i} = \emptyset \end{equation}</p>
<p>so .</p>
<h2>An orthogonal function space:</h2>
<p>The reason for is that each in the latent space of has compact support:</p>
<p>\begin{equation} \phi_i(X \setminus X_i) = 0 \end{equation}</p>
<p>which implies that we have a constant :</p>
<p>\begin{equation} \forall i, f_i(X \setminus X_i) = K \end{equation}</p>
<p>and therefore if we define :</p>
<p>\begin{equation} \forall g_i, g_{j \neq i} \in F_\theta-K, \langle g_i, g_j \rangle = \int_{x \in X} g_i(x)g_j(x) dx = 0 \end{equation}</p>
<p>so we may represent as follows:</p>
<p>\begin{equation} F_\theta(x) = \sum_i f(\phi_i(x)) + K = \sum_i f_i(x) + K \end{equation}</p>
<p>From this we may deduce that a deep rectifier network finds structure in by de-correlating the signals .</p>
<h1>An interpretation of Bernoulli dropout from a function space point of view:</h1>
<p>Given the above interpretation of deep rectifier networks, we may conjecture that Bernoulli dropout with probability half over each node serves as a method to maximise the effective size of as a function space so that . Mathematically, this argument may be justified by observing that Bernoulli dropout with probability half on each of the nodes is equivalent to sampling uniformly from a space of functions upon each function call during training.</p>
<p>We may evaluate this conjecture by checking whether, given a task such as MNIST classification, the usage of Bernoulli dropout increases the size of in terms of the number of orthogonal functions in as opposed to Bernoulli dropout with probability zero(i.e. no dropout). This number of functions can be approximated during the testing phase as done in (4) by applying a binary mask to ReLU activations, concatenating the binary activations so we have a binary vector per test datum and then counting the number of unique binary vectors.</p>
<p>One perspective that I believe will emerge naturally from this experimental analysis is that we will be able to investigate the inequality:</p>
<p>\begin{equation} H(X) \geq H(F_\theta(X)) \end{equation}</p>
<p>where is the Shannon Entropy and meausures the number of distinct functions .</p>
<h1>Discussion:</h1>
<p>At this point the reader may ask why might an ensemble of linear feature maps be good at approximating a mapping between two known unknowns, and where is a natural signal of some kind. My intuition as to why this is useful is that most(all?) natural signals are embedded on a manifold with a locally linear geometry. On this point, I have made a non-exhaustive list of references(5,6,7) to be consulted in the near future.</p>
<p>Finally, the reader may ask whether I have not made a mistake in saying that the functions have compact support. Surely, we have a linear projection of the domain into half-spaces. Mathematically, this is correct and this is something that I would say. However, it&rsquo;s important to note that all natural signals have a bounded range. The speed of electromagnetic waves in space is bounded by 3.10^8 m/s for example. Any values outside this range would be due to instrumentation error or represent a significant scientific discovery. For this reason I would argue that the compact support description is correct.</p>
<h1>References:</h1>
<ol>
<li>Nair, V. &amp; Hinton, G. Rectifief linear Units Improve Restricted Boltzmann Machines. 2010</li>
<li>Montufar, G. et al. On the Number of linear Regions of Deep Neural Networks. 2014.</li>
<li>Srivastava, N. et al. Dropout: A Simple Way to Prevent Neural Networks from Overfitting. 2014.</li>
<li>Srivastava R. et al. Understanding Locally Competitive Networks. 2014.</li>
<li>Sharpee T. , Rust N. &amp; Bialek W. Analyzing Neural Responses to Natural Signals: Maximally Informative Dimensions. 2004.</li>
<li>Peyr&eacute; G. Manifold Models for Signals and Images. 2009.</li>
<li>Schwartz O. &amp; Simoncelli E. Natural signal statistics and sensory gain control. 2001.</li>
</ol>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via Kepler Lounge<br>
<br>
March 1, 2019 at 06:59PM
